File: server/src/modules/revisedHeatmap.ts
import type { HeatmapRow, OTRecommendation } from '../types.js';

const ALL_INTERVALS = Array.from({ length: 48 }, (_, i) => {
  const h = String(Math.floor(i / 2)).padStart(2, '0');
  const m = i % 2 === 0 ? '00' : '30';
  return `${h}:${m}`;
});

function intervalIndex(time: string): number {
  const [h, m] = time.split(':').map(Number);
  return h * 2 + (m >= 30 ? 1 : 0);
}

/**
 * Computes a revised heatmap by adding OT headcount back to the intervals
 * covered by each OT recommendation.
 *
 * For each recommendation:
 * - Parse the OT time window (e.g. "05:00-07:00") or handle "Full Day"
 * - Add +1 to each half-hour interval in that window for that date/program
 */
export function computeRevisedHeatmap(
  originalData: HeatmapRow[],
  recommendations: OTRecommendation[],
): HeatmapRow[] {
  // Build a deltaMap that only tracks the +1 adjustments from OT recommendations
  const deltaMap = new Map<string, number>();

  for (const rec of recommendations) {
    const timeWindow = rec.otTimeWindow;
    let startIdx: number;
    let endIdx: number;

    if (timeWindow === 'Full Day') {
      startIdx = 0;
      endIdx = 48;
    } else {
      // Parse time window like "05:00-07:00" or "07:00-16:00"
      const match = /^(\d{2}:\d{2})-(\d{2}:\d{2})$/.exec(timeWindow);
      if (!match) continue;

      startIdx = intervalIndex(match[1]);
      endIdx = intervalIndex(match[2]);

      // Handle overnight shifts (e.g. 23:00-08:00)
      if (endIdx <= startIdx) endIdx += 48;
    }

    for (let i = startIdx; i < endIdx; i++) {
      const actualIdx = i % 48;
      const interval = ALL_INTERVALS[actualIdx];
      const key = `${rec.date}|${rec.program}|${interval}`;
      deltaMap.set(key, (deltaMap.get(key) ?? 0) + 1);
    }
  }

  // Apply deltas to original data — each row uses its own original value as the base
  return originalData.map((row) => {
    const key = `${row.date}|${row.program}|${row.intervalStartTime}`;
    const delta = deltaMap.get(key) ?? 0;
    return {
      ...row,
      overUnderValue: row.overUnderValue + delta,
    };
  });
}

File: python-server/modules/revised_heatmap.py
"""
Revised heatmap computation.
Mirrors revisedHeatmap.ts exactly.
"""
import re

ALL_INTERVALS = []
for i in range(48):
    h = str(i // 2).zfill(2)
    m = '00' if i % 2 == 0 else '30'
    ALL_INTERVALS.append(f"{h}:{m}")


def _interval_index(time_str):
    parts = time_str.split(':')
    h = int(parts[0])
    m = int(parts[1])
    return h * 2 + (1 if m >= 30 else 0)


def compute_revised_heatmap(original_data, recommendations):
    """Computes a revised heatmap by adding OT headcount back to intervals
    covered by each OT recommendation.

    Uses a delta-based approach: builds a delta_map of adjustments from
    recommendations, then applies them to the original data. Handles
    'Full Day' time windows by covering all 48 intervals.
    """
    delta_map = {}

    for rec in recommendations:
        time_window = rec.get('otTimeWindow', '')

        if time_window == 'Full Day':
            start_idx = 0
            end_idx = 48
        else:
            match = re.match(r'^(\d{2}:\d{2})-(\d{2}:\d{2})$', time_window)
            if not match:
                continue
            start_idx = _interval_index(match.group(1))
            end_idx = _interval_index(match.group(2))
            if end_idx <= start_idx:
                end_idx += 48

        for i in range(start_idx, end_idx):
            actual_idx = i % 48
            interval = ALL_INTERVALS[actual_idx]
            key = f"{rec['date']}|{rec['program']}|{interval}"
            delta_map[key] = delta_map.get(key, 0) + 1

    result = []
    for row in original_data:
        key = f"{row['date']}|{row['program']}|{row['intervalStartTime']}"
        delta = delta_map.get(key, 0)
        new_row = dict(row)
        new_row['overUnderValue'] = row['overUnderValue'] + delta
        result.append(new_row)
    return result
